### PR TITLE
[engine] create TX authority nodes from PTX

### PR DESCRIPTION
**Pull request by nickchapman-da**
Advances: #15882
- Create tx `Node.Authority` nodes in `endWithAuthority: PartialTransaction`,
- extend `foreachInExecutionOrder` and `foldInExecutionOrder` with new callbacks: `authorityBegin` `authorityEnd`, and fix up all callers
- Extend `calculateBlindingInfo`
- Improve testing in `WithAuthorityTest.scala` to check the constructed TXs have the expected shape.

Question: Should the new node type be regarded as an _Action_ node, and have `informeesOfNode: Set[Party]` ?

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 1]
```diff
@@ -661,14 +658,13 @@ private[speedy] case class PartialTransaction(
...
+        val authNode = Node.Authority(
+          obtained = info.authorizers,
+          children = context.children.toImmArray,
+        )
         copy(
-          context = info.parent.copy(
-            children = info.parent.children :++ context.children.toImmArray,
-            nextActionChildIdx = context.nextActionChildIdx,
-          )
+          context = info.parent.addActionChild(info.nodeId, context.minChildVersion),
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:666]: Why do we handle this node differently that Rollback ? 
(1) authority node are not an action node, and (2) they do not have version.

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:666]: So we should use `addRollbackChild` (perhaps renamed as `addNonActionChild`)?

````
    def addActionChild(child: NodeId, version: TxVersion): Context = {
      Context(info, minChildVersion min version, children :+ child, nextActionChildIdx + 1)
    }
    def addRollbackChild(child: NodeId, version: TxVersion, nextActionChildIdx: Int): Context =
      Context(info, minChildVersion min version, children :+ child, nextActionChildIdx)
```

Question: Why are we tracking `nextActionChildIdx: Int` in `Context` anyway?
Where is it used?

**Comment by remyhaemmerle-da** [comment]
> Question: Should the new node type be regarded as an Action node, and have informeesOfNode: Set[Party] ?

I will argue we should not bother too much with the design of the new node, as we will have to sell it to @andreaslochbihler-da when is back or remove it.

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM

**Pull request merged**